src/modm/processing/fiber/fiber.hpp

```diff
@@ -0,0 +1,151 @@
+
+#pragma once
+#include "context.h"
+#include <memory>
+
+namespace modm
+{
+
+class Fiber;
+
+namespace fiber
+{
+
+void yield();
+
+class Scheduler;
+class Waitable;
+template<class Data_t>
+class Channel;
+
+/**
+ * Stack captures a memory area used as fiber stack.
+ *
+ * @author	Erik Henriksson
+ * @ingroup	modm_processing_fiber
+ */
+template<size_t Size>
+class Stack
+{
+	static_assert(Size % 4 == 0, "Stack size must be multiple of 4.");
+	static_assert(Size >= 15*sizeof(void*), "Stack size must be ≥60B!");
+	// static_assert(Size >= 13*sizeof(void*), "Stack size must be ≥52B!");
+
+	friend class ::modm::Fiber;
+	friend class Scheduler;
+
+public:
+	constexpr Stack() = default;
+	Stack(const Stack&) = delete;
+
+private:
+	static constexpr size_t size = Size;
+	modm_aligned(8) uintptr_t memory[size / sizeof(uintptr_t)];
+};
+
+} // namespace fiber
+
+/**
+ * Fiber is a cooperative threading model consisting of an execution context (stack) and a
+ * function pointer representing the entry function.
+ *
+ * Fibers are scheduled in a round-robin fashion.
+ *
+ * @author Erik Henriksson
+ * @author Niklas Hauser
+ * @ingroup	modm_processing_fiber
+ */
+class Fiber
+{
+	friend class fiber::Scheduler;
+	template<class>
+	friend class fiber::Channel;
+	friend class fiber::Waitable;
+	friend void fiber::yield();
+
+public:
+	template<size_t Size, class T>
+	Fiber(fiber::Stack<Size>& stack, T&& closure);
+
+	template<size_t Size>
+	Fiber(fiber::Stack<Size>& stack, void(*fn)());
+
+protected:
+	inline void
+	jump(Fiber& other);
+
+private:
+	Fiber() = default;
+	Fiber(const Fiber&) = delete;
+
+	// Removes the current fiber from the scheduler and yields execution. This is called on fiber
+	// return.
+	static inline void
+	done();
+
+private:
+	modm_context_t ctx;
+	Fiber* next;
+};
+
+}	// namespace modm
+
+#include "scheduler.hpp"
+
+namespace modm
+{
+
+template<size_t Size>
+Fiber::Fiber(fiber::Stack<Size>& stack, void(*fn)())
+{
+	ctx = modm_context_init((uintptr_t) stack.memory,
+	                        ((uintptr_t) stack.memory) + stack.size,
+	                        (uintptr_t) fn,
+	                        (uintptr_t) fiber::Scheduler::deregisterFiber);
+	// register this fiber to be scheduled
+	fiber::Scheduler::registerFiber(this);
+}
+
+template<size_t Size, class T>
+Fiber::Fiber(fiber::Stack<Size>& stack, T&& closure)
+{
+	// Find a suitable aligned area at the top of stack to allocate the closure
+	uintptr_t ptr = uintptr_t(stack.memory) + stack.size;
+	if constexpr (sizeof(std::decay_t<T>))
+	{
+		ptr -= sizeof(std::decay_t<T>);
+		ptr &= ~(std::max(sizeof(void*), alignof(std::decay_t<T>)) - 1u);
+		// construct closure in place
+		::new ((void*)ptr) std::decay_t<T>{std::forward<T>(closure)};
+	}
+	// get the function pointer from the closure via this union
+	union {
+		decltype(&std::decay_t<T>::operator()) lambda = &std::decay_t<T>::operator();
+		uintptr_t function;
+	};
```

Comment:
There is a pretty easy solution for handling the member function pointer ABI madness. Let the compiler generate the code for you:
```suggestion
        auto function = (uintptr_t) +[](std::decay_t<T>* closure) {
                (*closure)();
        };
```
This will work for the pointer to virtual function case.

Reply:
Yup, that works great, simple and very nice, love it!